sky/backends/backend_utils.py

```diff
@@ -1132,14 +1137,67 @@ def query_head_ip_with_retries(cluster_yaml: str, max_attempts: int = 1) -> str:
     return head_ip
 
 
+@timeline.event
+def get_stable_cluster_ips(handle: backends.Backend.ResourceHandle,
+                           cluster_external_ips: Optional[List[str]] = None,
+                           cluster_internal_ips: Optional[List[str]] = None,
+                           get_internal_ips: bool = False,
+                           head_ip_max_attempts: int = 1,
+                           worker_ip_max_attempts: int = 1,
+                           use_cached_ips: bool = True) -> List[str]:
+    """Returns a stable list of (internal IP, external IP) tuples.
+
+    The lists of cluster IPs that are passed must correspond to each other.
+    """
+    if use_cached_ips:
+        if get_internal_ips and handle.internal_ips is not None:
+            return handle.internal_ips
+        elif handle.external_ips is not None:
+            return handle.external_ips
+
+    if cluster_external_ips is None:
+        cluster_external_ips = get_node_ips(
+            handle.cluster_yaml,
+            handle.launched_nodes,
+            handle=handle,
+            head_ip_max_attempts=head_ip_max_attempts,
+            worker_ip_max_attempts=worker_ip_max_attempts,
+            get_internal_ips=False)
+    if cluster_internal_ips is None:
+        cluster_internal_ips = get_node_ips(
+            handle.cluster_yaml,
+            handle.launched_nodes,
+            handle=handle,
+            head_ip_max_attempts=head_ip_max_attempts,
+            worker_ip_max_attempts=worker_ip_max_attempts,
+            get_internal_ips=True)
+
+    assert len(cluster_external_ips) == len(
+        cluster_internal_ips
+    ), 'length of external IPs and internal IPs must be the same'
+
+    internal_external_ips = list(zip(cluster_internal_ips,
+                                     cluster_external_ips))
+
+    # Ensure head node is the first element, then sort based on the external
+    # IPs for stableness
+    stable_internal_external_ips = [internal_external_ips[0]] + sorted(
+        internal_external_ips[1:], key=lambda x: x[1])
+    handle.stable_internal_external_ips = stable_internal_external_ips
+    if get_internal_ips:
+        return handle.internal_ips
+    elif handle.external_ips:
+        return handle.external_ips
+
+
 @timeline.event
 def get_node_ips(cluster_yaml: str,
                  expected_num_nodes: int,
                  handle: Optional[backends.Backend.ResourceHandle] = None,
                  head_ip_max_attempts: int = 1,
                  worker_ip_max_attempts: int = 1,
                  get_internal_ips: bool = False) -> List[str]:
-    """Returns the IPs of all nodes in the cluster."""
+    """Returns the IPs of all nodes in the cluster, with head node at front."""
```

Comment:
The L1212-L1214 seems not needed now, since the caching is mechanism is already handled by the `get_stable_cluster_ips()`